ONNX Object Detection Tutorial (#12755)
dotnet/docs: docs/machine-learning/tutorials/object-detection-onnx.md

---
title: 'Tutorial: Detect objects using deep learning with ONNX and ML.NET'
description: This tutorial illustrates how to use a pre-trained ONNX deep learning model in ML.NET to detect objects in images.
author: luisquintanilla
ms.author: luquinta
ms.date: 07/10/2019
ms.topic: tutorial
ms.custom: mvc
#Customer intent: As a developer, I want to use ML.NET so that I can use a pre-trained model in an object detection scenario to detect objects in images using ONNX.
---

# Tutorial: Detect objects using ONNX in ML.NET

Learn how to use a pre-trained ONNX model in ML.NET to detect objects. The pre-trained model was trained to find and classify objects in an image.

Training an object detection model from scratch requires setting millions of parameters, a large amount of labeled training data and a vast amount of compute resources (hundreds of GPU hours). Using a pre-trained model allows you to shortcut the training process.

In this tutorial, you learn how to:
> [!div class="checklist"]

> > * Understand the problem

Reviewer: I think we could expand this list to make it a little less generic. They learn a lot in this tutorial!

This sample creates a .NET core console application that detects objects within an image using a pre-trained deep learning ONNX model. The code for this sample can be found on the [dotnet/machinelearning-samples repository](https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/getting-started/DeepLearning_ObjectDetection_Onnx) on GitHub.

## What is object detection

Object detection is a computer vision problem. While closely related to image classification, object detection performs image classification at a more granular scale. When an image is processed, there is first a check on whether there is an object present or not. If so, the detected object is classified and using object localization, the location of that object is returned along with the class. Typically, you want to use this approach on images where there are multiple objects of different types so that each is individually recognized. 

![](./media/object-detection-onnx/img-classification-obj-detection.PNG)

Some use cases for object detection include:

- Self-Driving Cars
- Robotics
- Face Detection
- Workplace Safety
- Object Counting
- Activity Recognition

> ## Select the appropriate machine learning task 

Reviewer: I'm still of the opinion that this title doesn't describe the content of this section

Object detection is an image processing task. Therefore, most deep learning models trained to solve this problem are CNNs. The model used in this tutorial is the Tiny YOLOv2 model, a more compact version of the YOLOv2 model described in the paper: ["YOLO9000: Better, Faster, Stronger" by Redmon and Fadhari](https://arxiv.org/pdf/1612.08242.pdf). Tiny YOLOv2 is trained on the Pascal VOC dataset and is made up of 15 layers that can predict 20 different classes of objects. Because Tiny YOLOv2 is a condensed version of the original YOLOv2 model, a tradeoff is made between speed and accuracy. 

The model is stored in ONNX format. The Open Neural Network Exchange (ONNX) is an open source format for AI models. ONNX supports interoperability between frameworks. This means you can train a model in one of the many popular machine learning frameworks like PyTorch, convert it into ONNX format and consume the ONNX model in a different framework like ML.NET. To learn more, visit the [ONNX website](https://onnx.ai/). 

> ![](./media/object-detection-onnx/onnx-frameworks.png)

Reviewer: A diagram with the network inputs and outputs would be very useful here. If you draft it, we can ask Deborah in art to make it professional

> ## Create console application

Reviewer: Is this the best title for this section? Can we make it more specific?

    - `Label` contains the name of the file.

    Additionally, `ImageData` contains a method `ReadFromFile` which loads multiple image files stored in the `imageFolder` path specified and returns them as a collection of `ImageNetData` objects.

Create your prediction class in the *DataStructures* directory.

1. In **Solution Explorer**, right-click the *DataStructures* directory, and then select **Add** > **New Item**.
1. In the **Add New Item** dialog box, select **Class** and change the **Name** field to *ImageNetPrediction.cs*. Then, select the **Add** button.

    The *ImageNetPrediction.cs* file opens in the code editor. Add the following `using` statement to the top of *ImageNetPrediction.cs*:

    ```csharp
    using Microsoft.ML.Data;
    ```

    Remove the existing class definition and add the following code for the `ImageNetPrediction` class to the *ImageNetPrediction.cs* file:

    ```csharp
    public class ImageNetPrediction
    {
        [ColumnName("grid")]
        public float[] PredictedLabels;
    }
    ```

    `ImageNetPrediction` is the prediction data class and has the following `float[]` field:

>     - `PredictedLabel` contains the dimensions, objectness score and class probabilities for each of the bounding boxes detected in an image.

Author: That's what it's referred to as in the paper.

https://arxiv.org/pdf/1612.08242.pdf

thumbnailGraphic.DrawRectangle(pen, x, y, width, height);
```

Outside of the for-each loop, add code to save the images in the `outputDirectory`.

```csharp
if (!Directory.Exists(outputImageLocation))
{
    Directory.CreateDirectory(outputImageLocation);
}

image.Save(Path.Combine(outputImageLocation, imageName));
```

To get additional feedback that the application is making predictions as expected at runtime, add a method called `LogDetectedObjects` below the `DrawBoundingBox` method in the *Program.cs* file to output the detected objects to the console.

```csharp
private static void LogDetectedObjects(string imageName, IList<YoloBoundingBox> boundingBoxes)
{
    Console.WriteLine($".....The objects in the image {imageName} are detected as below....");

    foreach (var box in boundingBoxes)
    {
        Console.WriteLine($"{box.Label} and its Confidence score: {box.Confidence}");
    }

    Console.WriteLine("");
}
```

Both of these methods will be useful when the model has produced outputs and those have been processed. First though, create the functionality to process the model outputs.

## Create a parser to post-process model outputs

> The expected output generated by the pre-trained ONNX model is a tensor of shape `125 x 13 x 13` that is returned as a one-dimensional `float[]` of 21125 elements. In order to extract the predictions generated by the model, some post-processing work is required. To do so, create a set of classes to help parse the output.

Reviewer: Should we define what a tensor is?

Reviewer: A diagram showing the mapping would be useful here

---
title: 'Tutorial: Detect objects using deep learning with ONNX and ML.NET'
description: This tutorial illustrates how to use a pre-trained ONNX deep learning model in ML.NET to detect objects in images.
author: luisquintanilla
ms.author: luquinta
ms.date: 07/29/2019
ms.topic: tutorial
ms.custom: mvc
#Customer intent: As a developer, I want to use ML.NET so that I can use a pre-trained model in an object detection scenario to detect objects in images using ONNX.
---

# Tutorial: Detect objects using ONNX in ML.NET

> Learn how to use a pre-trained ONNX model in ML.NET to detect objects. The pre-trained model was trained to find and classify objects in an image.

Reviewer: Suggest simplifying this to: 
```
Learn how to use a pre-trained ONNX model in ML.NET to detect objects in images
```

> * Detect objects with a loaded model

## Pre-requisites

- [Visual Studio 2017 15.6 or later](https://visualstudio.microsoft.com/downloads/?utm_medium=microsoft&utm_source=docs.microsoft.com&utm_campaign=inline+link&utm_content=download+vs2017) with the ".NET Core cross-platform development" workload installed.
- [Microsoft.ML NuGet Package](https://www.nuget.org/packages/Microsoft.ML/)
- [Microsoft.ML.ImageAnalytics NuGet Package](https://www.nuget.org/packages/Microsoft.ML.ImageAnalytics/)
- [Microsoft.ML.OnnxTransformer NuGet Package](https://www.nuget.org/packages/Microsoft.ML.OnnxTransformer/)
- [Tiny YOLOv2 pre-trained model](https://github.com/onnx/models/tree/master/tiny_yolov2)
- [Netron](https://github.com/lutzroeder/netron) (optional)

## ONNX object detection sample overview

This sample creates a .NET core console application that detects objects within an image using a pre-trained deep learning ONNX model. The code for this sample can be found on the [dotnet/machinelearning-samples repository](https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/getting-started/DeepLearning_ObjectDetection_Onnx) on GitHub.

## What is object detection?

> Object detection is a computer vision problem. While closely related to image classification, object detection performs image classification at a more granular scale. When an image is processed, there's first a check on whether there's an object present or not. If so, the detected object is classified and using object localization, the location of that object is returned along with the class. Typically, you want to use this approach on images where there are multiple objects of different types so that each is individually recognized. 

Reviewer: Suggest simplifying to:
`Object detection both locates _and_ categorizes entities within images. Use object detection when images contain multiple objects of different types.`

![](./media/object-detection-onnx/img-classification-obj-detection.PNG)

Some use cases for object detection include:

- Self-Driving Cars
- Robotics
- Face Detection
- Workplace Safety
- Object Counting
- Activity Recognition

## Select the appropriate machine learning task 

Deep learning is a subset of machine learning. To train deep learning models, large quantities of data are required. Patterns in the data are represented by a series of layers. The relationships in the data are encoded as connections between the layers containing weights. The higher the weight, the stronger the relationship. Collectively, this series of layers and connections are known as artificial neural networks. The more layers in a network, the "deeper" it is, making it a deep neural network. 

> There are different types of neural networks, each more suitable than the other for certain tasks. The most common types of neural networks are Multi-Layered Perceptron (MLP), Convolutional Neural Network (CNN) and Recurrent Neural Network (RNN). The most basic is the MLP, which maps a set of inputs to a set of outputs. This neural network is good when the data does not have a spatial or time component. The CNN as implied by the name makes use of convolutional layers that take into account spatial information contained in the data. A good use case for CNNs is image processing when trying to detect the presence of a particular feature in a region of an image (i.e. is there a nose in the center of an image?). Finally, RNNs allow for the persistence of state or memory to be used as input therefore making it a good candidate for time-series analysis where the sequential ordering and context of events is important. 

Reviewer: ```suggestion
There are different types of neural networks, the most common being Multi-Layered Perceptron (MLP), Convolutional Neural Network (CNN) and Recurrent Neural Network (RNN). The most basic is the MLP, which maps a set of inputs to a set of outputs. This neural network is good when the data does not have a spatial or time component. The CNN makes use of convolutional layers to process spatial information contained in the data. A good use case for CNNs is image processing to detect the presence of a feature in a region of an image (for example, is there a nose in the center of an image?). Finally, RNNs allow for the persistence of state or memory to be used as input. RNNs are used for time-series analysis, where the sequential ordering and context of events is important. 
```

![](./media/object-detection-onnx/netron-model-map.png)

> The YOLO model takes an image `3(RGB) x 416px x 416px`. The model takes this input and passes it through the different layers to produce an output. The output divides the input image into a `13 x 13` grid, with each box in the grid consisting of `125` values. 

Reviewer: ```suggestion
The YOLO model takes an image `3(RGB) x 416px x 416px`. The model takes this input and passes it through the different layers to produce an output. The output divides the input image into a `13 x 13` grid, with each cell in the grid consisting of `125` values. 
```

> ![](./media/object-detection-onnx/netron-model-map.png)

Reviewer: The netron part of this image is too small. I can help rework this image

> The input expected by the YOLO model is an image that is `416px x 416px`, where each pixel contains 3 data elements corresponding to the RGB values of the respective pixel. The model takes this input and passes it through the different layers to produce an output. The output is a grid of dimensions `125 x 13 x 13`. 

Reviewer: ```suggestion
The YOLO model takes an image `416px x 416px x 3 (RGB)`. The model takes this input and passes it through the different layers to produce an output. The output divides the input image into a `13 x 13` grid, with each box in the grid consisting of `125` values. 
```

> ![](./media/object-detection-onnx/model-output-description.png)

Reviewer: Image looks good! Would the yellow bounding box be where the centroid of the object is though?

> The model segments an image into a `13 x 13` grid, where each grid cell is `32px x 32px`. Each grid cell contains 125 elements within it that correspond to 5 areas in an image which are believed to contain an object, also known as bounding boxes. Each of these bounding boxes stores 25 elements of information to describe itself:

Reviewer: ```suggestion
The model segments an image into a `13 x 13` grid, where each grid cell is `32px x 32px`. Each grid cell contains 5 potential object bounding boxes. A bounding box has  25 elements:
```

Create your prediction class in the *DataStructures* directory.

1. In **Solution Explorer**, right-click the *DataStructures* directory, and then select **Add** > **New Item**.
1. In the **Add New Item** dialog box, select **Class** and change the **Name** field to *ImageNetPrediction.cs*. Then, select the **Add** button.

    The *ImageNetPrediction.cs* file opens in the code editor. Add the following `using` statement to the top of *ImageNetPrediction.cs*:

    [!code-csharp [ImageNetPredictionUsings](~/machinelearning-samples/samples/csharp/getting-started/DeepLearning_ObjectDetection_Onnx/ObjectDetectionConsoleApp/DataStructures/ImageNetPrediction.cs#L1)]

    Remove the existing class definition and add the following code for the `ImageNetPrediction` class to the *ImageNetPrediction.cs* file:

    [!code-csharp [ImageNetPredictionClass](~/machinelearning-samples/samples/csharp/getting-started/DeepLearning_ObjectDetection_Onnx/ObjectDetectionConsoleApp/DataStructures/ImageNetPrediction.cs#L5-L9)]

>     `ImageNetPrediction` is the prediction data class and has the following `float[]` field:

Reviewer: I suggest moving the grid image into this section, to break up the code a bit, and also to illustrate how the ImagePrediction data is encoded.

We could talk about the expected dimensions of the ImagePrediction float[] array too

[!code-csharp [LogOuptuts](~/machinelearning-samples/samples/csharp/getting-started/DeepLearning_ObjectDetection_Onnx/ObjectDetectionConsoleApp/Program.cs#L133-L143)]

Both of these methods will be useful when the model has produced outputs and those have been processed. First though, create the functionality to process the model outputs.

## Create a parser to post-process model outputs

The output generated by the pre-trained ONNX model is a float array of length 21125, representing the elements of a tensor with dimensions `125 x 13 x 13`. In order to transform the predictions generated by the model into a tensor, some post-processing work is required. To do so, create a set of classes to help parse the output.

Add a new directory to your project to organize the set of parser classes.

1. In **Solution Explorer**, right-click the project, and then select **Add** > **New Folder**. When the new folder appears in the Solution Explorer, name it "YoloParser".

### Create bounding boxes and dimensions

> Object localization is a core component of object detection. As such, much of the data output by the model contains coordinates and dimensions of the bounding boxes of objects within the image. Create a base class for dimensions.

Reviewer: ```suggestion
The data output by the model contains coordinates and dimensions of the bounding boxes of objects within the image. Create a base class for dimensions.
```
